cpp/dolfinx/function/Function.h

```diff
@@ -89,7 +95,21 @@ class Function
   }
 
   /// Move assignment
-  Function& operator=(Function&& v) = default;
+  Function& operator=(Function&& v)
+  {
+    if (this != &v)
+    {
+      name = std::move(v.name);
+      _id = std::move(v._id);
+      _function_space = std::move(v._function_space);
+      _x = std::move(v._x);
+      if (_petsc_vector)
```

Comment:
```
if (_petsc_vector)
        VecDestroy(&_petsc_vector);
```
isn't required. Ownership is just being transferred.

Reply:
Shouldn't we clean up the memory the destination function held before the assignment?

Reply:
I mean, in case we have something like

```
// Both functions store data
u = function::Function<PetscScalar>(V);
v = function::Function<PetscScalar>(V);

u = std::move(v);
```
Then we should free the memory pointed to by `u._petsc_vector` before acquiring ownership.

Reply:
Isn't this dealt with by the destructor and the RAII idiom in the usual way?

Reply:
That data member is stored as a plain pointer.

Reply:
Yes, it should be cleaned up by RAII in the `Function` destructor. What needs to be done here is `v._petsc_vector` needs to be set to `nullptr` is there is nothing to cleanup. The simplest fix is probably to use `std::swap` (or remove PETSc!).

Reply:
I verified with valgrind that there is, in fact, a memory leak without a call to `VecDestroy`.

Reply:
I think @massimiliano-leoni is right, after `u = std::move(v)` the old `u._petsc_vector` got lost and no-one is responsible for it.

Reply:
What's needed is replacing `std::exchange` with `std::swap`.

Reply:
So you suggest to swap `u._petsc_vector` and `v._petsc_vector` such that after `u = std::move(v)` the `v` will be managing the old u’s vector? That would work as well to avoid leaks, but I do not undertstand why bind these two together. I would expect after move assigning `v` into `u` the old u’s vector be freed. It could even happen that someone stores a reference to `v` so it won’t get freed.

Reply:
Amazing @massimiliano-leoni, I learned something new. Thanks for running valgrind on it. C++ move semantics are so transparent and easy to understand ;).

Reply:
@michalhabera a couple of points why destruction should be left the destructor:
1. The move assignment should just 'move', i.e. be essentially zero ops. Cleaning up memory in a move assignment is counter-intuitive
2. Destruction should take place in the destructor
3. Adding `VecDestroy` to the move assignment is code duplication - the step is already performed by `Function::~Function`

Reply:
@massimiliano-leoni Giving it some thoughts I agree with Garth on doing `std::swap(u._petsc_vector, v._petsc_vector)` and not calling `VecDestroy` inside the move assignment.

Reply:
@jhale I know the feeling!

@garth-wells , @michalhabera there are notable proponents of both implementations, and I'm fine with either, so I changed it to the one you prefer.